File: src/org/onion_lang/onion/tools/option/CommandLineParser.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package org.onion_lang.onion.tools.option;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Kota Mizushima
 * Date: 2005/04/08
 */
public class CommandLineParser {
  private final OptionConf[] confs;
  
  public CommandLineParser(OptionConf[] confs) {
    this.confs = confs;
  }
  
  public ParseResult parse(String[] cmdline){
    Map<String, Object> noArgOpts = new HashMap<String, Object>();
    Map<String, String> opts = new HashMap<String, String>();
    List<String> args = new ArrayList<String>();
    List<String> lackedOptNames = new ArrayList<String>();
    List<String> invalidOptNames = new ArrayList<String>();
    {
      int i;
      for(i = 0; i < cmdline.length && cmdline[i].startsWith("-");) {
        String param = cmdline[i];
        OptionConf conf = getConf(param);
        if(conf == null){
          invalidOptNames.add(param);
          i++;
        }else if(conf.hasArgument()){
          if(i + 1 >= cmdline.length){
            lackedOptNames.add(param);
          }else{
            opts.put(param, cmdline[i + 1]);
          }
          i+=2;
        }else{
          noArgOpts.put(param, new Object());
          i++;
        }
      }
      for(; i < cmdline.length; i++) {
        args.add(cmdline[i]);
      }
    }
    if(lackedOptNames.size() == 0 && invalidOptNames.size() == 0){
      return new ParseSuccess(noArgOpts, opts, args);
    }else{
      return new ParseFailure(
        lackedOptNames.toArray(new String[0]), 
        invalidOptNames.toArray(new String[0]));
    }
  }
  
  private OptionConf getConf(String optionName){
    for(int i = 0; i < confs.length; i++){
      if(confs[i].getOptionName().equals(optionName)){
        return confs[i];
      }
    }
    return null;
  }
}

File: src/main/scala/onion/tools/option/CommandLineParser.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005-2012, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.tools.option

import scala.collection.mutable

/**
 * @author Kota Mizushima
 *
 */
class CommandLineParser(val configs: OptionConfig*) {

  def parse(cmdline: Array[String]): ParseResult = {
    val opts = mutable.Map[String, CommandLineParam]()
    val args = mutable.Buffer[String]()
    val lackedOptNames = mutable.Buffer[ValuedParam]()
    val invalidOptNames = mutable.Buffer[ValuedParam]()
    var i = 0

    while (i < cmdline.length) {
      if (cmdline(i).startsWith("-")) {
        val param = cmdline(i)
        val config: Option[OptionConfig] = configs.find(_.optionName == param)
        config match {
          case None =>
            invalidOptNames += ValuedParam(param)
            i += 1
          case Some(config) =>
            if (config.hasArgument) {
              if (i + 1 >= cmdline.length) {
                lackedOptNames += ValuedParam(param)
              } else {
                opts(param) = ValuedParam(cmdline(i + 1))
              }
              i += 2
            } else {
              opts(param) = NoValuedParam
              i += 1
            }
        }
      } else {
        args += cmdline(i)
        i += 1
      }
    }

    if (lackedOptNames.size == 0 && invalidOptNames.size == 0)
      new ParseSuccess(opts, args.toArray)
    else
      new ParseFailure(lackedOptNames.toArray, invalidOptNames.toArray)
  }
}